### PR TITLE
SONARJAVA-3462 S4036: Searching OS commands in PATH is security-sensitive 

### PR DESCRIPTION
**Pull request by dorian-burihabwa-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree argument = arguments.get(0);
+    processArgument(argument);
+  }
+
+  private void processArgument(ExpressionTree tree) {
+    ExpressionTree argument = ExpressionUtils.skipParentheses(tree);
+    if (argument.is(Tree.Kind.STRING_LITERAL)) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:140]: Maybe a `switch` would be cleaner here.

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    Tree declaration = symbol.declaration();
+    if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+      return true;
+    }
+    VariableTree variable = (VariableTree) declaration;
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null) {
+      return true;
+    }
+    Tree.Kind initializerKind = initializer.kind();
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:185]: As discussed, you can extract this code into a separate method.

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    "../",
+    "~/",
+    "\\",
+    ".\\",
+    "..\\"
+  );
+
+  private static final Pattern WINDOWS_DISK_PATTERN = Pattern.compile("^[A-Z]:\\\\.*");
+
+  private static final String MESSAGE = "Make sure the \"PATH\" used to find this command includes only what you intend.";
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:101]: What do you think of `Make sure that the command can be found in the PATH.`

[the remark this answers is not in the document]
Comment (quentin-jaquier-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:101]: I was confused by the fact that PATH was between quotes, but when I re-read the description, it makes sense, maybe it's fine to keep it as it is.

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java (hunk 2 above)
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java (hunk 1 above)
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    return !command.isPresent() || isCompliant(command.get());
+  }
+
+  private static boolean isIdentifierCommandValid(IdentifierTree identifier) {
+    Symbol symbol = identifier.symbol();
+    Type type = symbol.type();
+    if (type.is(STRING_TYPE)) {
+      return isStringLiteralCommandValid(identifier);
+    }
+    //The identifier must be an array or a list
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+      return true;
+    }
+    VariableTree variable = (VariableTree) declaration;
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null) {
+      return true;
+    }
+    Tree.Kind initializerKind = initializer.kind();
+    if (initializerKind.equals(Tree.Kind.NEW_ARRAY)) {
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+      return isNewArrayCommandValid(initializer);
+    }
+    if (initializerKind.equals(Tree.Kind.METHOD_INVOCATION)) {
+      return isListCommandValid(initializer);
+    }
+    return true;
+  }
+
+  private static boolean isListCommandValid(ExpressionTree listInitializer) {
+    MethodInvocationTree invocation = (MethodInvocationTree) listInitializer;
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+      reportIssue(argument, MESSAGE);
+    } else if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+      if (isListCommandValid(argument)) {
+        return;
+      }
+      reportIssue(argument, MESSAGE);
+    }
+  }
+
+  private static boolean isStringLiteralCommandValid(ExpressionTree stringLitteral) {
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree tree) {
+    Arguments arguments = tree.arguments();
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree argument = arguments.get(0);
+    processArgument(argument);
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+      PROCESS_BUILDER_MATCHER,
+      PROCESS_BUILDER_COMMAND_MATCHER
+    );
+  }
+
+
+  private static boolean isCompliant(String command) {
+    return STARTS.stream().anyMatch(command::startsWith) ||
+      WINDOWS_DISK_PATTERN.matcher(command).matches();
+  }
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java]: Done

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        break;
+    }
+  }
+
+  private static boolean isCompliant(String command) {
+    return STARTS.stream().anyMatch(command::startsWith) ||
+      WINDOWS_DISK_PATTERN.matcher(command).matches();
+  }
+
+  private static boolean isStringLiteralCommandValid(ExpressionTree stringLiteral) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:162]: In fact, the argument is not always a "stringLiteral", it can be an identifier aswell. We should name it expression to better reflect that.

File: java-checks-test-sources/src/main/java/checks/OSCommandsPath.java [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+  private static final List<String> COMPLIANT_COMMAND_LIST_WINDOWS_PARENT = Arrays.asList("..\\make");
+  private static final List<String> COMPLIANT_COMMAND_LIST_WINDOWS_CURRENT = Arrays.asList(".\\make");
+  private static final List<String> COMPLIANT_COMMAND_LIST_WINDOWS_NETWORK = Arrays.asList("\\\\SERVER\\make");
+  private static final List<String> COMPLIANT_COMMAND_LIST_VARIABLE = Arrays.asList(File.pathSeparator);
+
+  private static final String UNINITIALIZED_COMMAND = null;
+  private static final String[] UNINITIALIZED_COMMAND_ARRAY = null;
+  private static final List<String> UNINITIALIZED_COMMAND_LIST = null;
+
+  public void falseNegatives() throws IOException {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks-test-sources/src/main/java/checks/OSCommandsPath.java:78]: I would rather put this method later in the file, because it would make more sense to have the cases where the rule reports an issue first, before having to understand the cases where the rule does not report an issue.

File: java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    Optional<String> command = stringLiteral.asConstant(String.class);
+    return !command.isPresent() || isCompliant(command.get());
+  }
+
+  private static boolean isIdentifierCommandValid(IdentifierTree identifier) {
+    Symbol symbol = identifier.symbol();
+    if (!symbol.isFinal()) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java:169]: Any reason why we don't check for effectively final variables here?

**Review by quentin-jaquier-sonarsource — COMMENTED** [review]
In addition, don't forget to test the issue location!

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
Implementation LGTM!

We are just missing a test when the initializer is null, to cover one additional line.
It's okay to not cover the others though.
